tests/infra/partitions.py

```diff
@@ -0,0 +1,196 @@
+        LOG.info(f"{CCF_IPTABLES_CHAIN} iptables chain cleaned up")
+
+    def __init__(self, network):
+        self.network = network
+        # Create iptables chain
+        if not iptc.easy.has_chain("filter", CCF_IPTABLES_CHAIN):
+            iptc.easy.add_chain("filter", CCF_IPTABLES_CHAIN)
+
+        # Create iptables rule in INPUT chain
+        if not iptc.easy.has_rule("filter", "INPUT", CCF_INPUT_RULE):
+            iptc.easy.insert_rule("filter", "INPUT", CCF_INPUT_RULE)
+
+    def isolate_node(
+        self,
+        node: infra.node.Node,
+        other: Optional[infra.node.Node] = None,
+    ):
+        """
+        Isolates a single :py:class:`infra.node.Node` from the network, or from a specific other node if specified.
+
+        :param infra.node.Node node: The :py:class:`infra.node.Node` to isolate.
+        :param Optional[infra.node.Node] other: The other node to isolate node from (optional).
+
+        :return: :py:class:`infra.partitions.Rules`
+        """
+        if node is other:
+            return None
+
+        base_rule = {"protocol": "tcp", "target": "DROP"}
+        name = f"Isolate node {node.local_node_id}"
+
+        # Isolates node server socket
+        server_rule = {
+            **base_rule,
+            "dst": node.host,
+            "tcp": {"dport": str(node.node_port)},
+        }
+
+        # Isolates all node client sockets
+        client_rule = {
+            **base_rule,
+            "src": node.node_client_host,
+        }
+
+        # If there is one, only isolate from specific node
+        if other:
+            server_rule["src"] = other.node_client_host
+            client_rule["dst"] = other.host
+            name += f" from node {other.local_node_id}"
+
+        if iptc.easy.has_rule("filter", CCF_IPTABLES_CHAIN, server_rule):
+            iptc.easy.delete_rule("filter", CCF_IPTABLES_CHAIN, server_rule)
+
+        if iptc.easy.has_rule("filter", CCF_IPTABLES_CHAIN, client_rule):
+            iptc.easy.delete_rule("filter", CCF_IPTABLES_CHAIN, client_rule)
+
+        iptc.easy.insert_rule("filter", CCF_IPTABLES_CHAIN, server_rule)
+        iptc.easy.insert_rule("filter", CCF_IPTABLES_CHAIN, client_rule)
+
+        return Rules([server_rule, client_rule], name)
+
+    @staticmethod
+    def _get_partition_name(partition: List[infra.node.Node]):
+        if not partition:
+            return ""
+        return f'[{",".join(str(node.local_node_id) for node in partition)}]'
+
+    def partition(
+        self,
+        *args: List[infra.node.Node],
+        **kwargs,
+    ):
+        """
+        Creates an arbitrary number of partitions of :py:class:`infra.node.Node`. All other joined nodes in the
+        :py:class:`infra.network.Network` are also isolated in their own partition.
+
+        :param List[infra.node.Node] *args: A variable length argument list of :py:class:`infra.node.Node` (i.e. partitions) to isolate.
+        :param str **name: Name of the partition (optional).
```

Comment:
What's the use case for setting a name here? It seems like we want the names to be canonical, so we can reliably detect when we already have a collision, so its risky for the caller to override this? Perhaps this name is not used for clearing things later.

If it needs to be canonical, I think the partitions should be specified as `set`s rather than `list`s, since the partitions `[n1, n2]` and `[n2, n1]` are logically equivalent and we want to give them the same name.

Reply:
As discussed, the name doesn't have to be canonical as the rules are already identified by their value (iptables rule dict). It is only optionally used to easily name/identify partition rules from the caller. I'm leaning towards leaving it in as our use case for testing network partitions isn't fully defined yet.